csharp/src/Google.Protobuf/Reflection/GeneratedClrTypeInfo.cs

```diff
@@ -88,7 +98,11 @@ public sealed class GeneratedClrTypeInfo
         /// Each array parameter may be null, to indicate a lack of values.
         /// The parameter order is designed to make it feasible to format the generated code readably.
         /// </summary>
-        public GeneratedClrTypeInfo(Type clrType, MessageParser parser, string[] propertyNames, string[] oneofNames, Type[] nestedEnums, Extension[] extensions, GeneratedClrTypeInfo[] nestedTypes)
+        public GeneratedClrTypeInfo(
+            // Preserve all public members on message types when trimming is enabled.
+            // This ensures that members used by reflection, e.g. JSON serialization, are preserved.
+            [DynamicallyAccessedMembers(MessageAccessibility)]
```

Comment:
> is this really the best place for placing the DynamicallyAccessedMembers attribute?

The idea of dynamically accessed members attribute is to annotate those places so `Type` arguments and generic arguments that are passed to them can be used to include members on those types, without having them explicitly referenced by code. Having `GeneratedClrTypeInfo` included in code generation seems like a good way to include members on those types that could be used dynamically.

> I can imagine there being use cases where one uses large .proto files (quite common e.g. at google) and only uses a subset of the fields in the message. So one use case to think about is whether trimming can be made to strip all the unused protobuf fields (assuming the user doesn't use reflection on json serialization). I think this is probably quite difficult to achieve, but is something to think about in the future. Also this should be listed as a "limitation" of the current approach?.

Including members on types like this could definitely pull in more than you need. There is a tradeoff between everything just working but the app size is larger, and some functionality breaking but smaller apps.

Right now I'm not exactly sure how granular and smart trimming is. For example, imagine a library assembly that has generated types from two large proto files. If your app doesn't use any types from one of the library's proto files, will its static ctor never be evaluated, and therefore `GeneratedClrTypeInfo` calls are trimmed and so none of its types will be included?